--- irc2.11.2p3.orig/common/struct_def.h	2009-11-13 21:08:10.000000000 +0100
+++ irc2.11.2p3/common/struct_def.h	2019-09-24 22:51:09.792535845 +0200
@@ -192,6 +192,7 @@
 #define FLAGS_RESTRICT	0x0010 /* restricted user */
 #define FLAGS_AWAY	0x0020 /* user is away */
 #define FLAGS_EXEMPT    0x0040 /* user is exempted from k-lines */
+#define FLAGS_SPOOFED   0x0080 /* user is spoofed */
 #ifdef XLINE
 #define FLAGS_XLINED	0x0100	/* X-lined client */
 #endif
@@ -220,6 +221,7 @@
 #define	IsDead(x)		((x)->flags & FLAGS_DEADSOCK)
 #define	IsBursting(x)		(!((x)->flags & FLAGS_EOB))
 #define IsKlineExempt(x)        ((x)->user && (x)->user->flags & FLAGS_EXEMPT)
+#define IsSpoofed(x)            ((x)->user && (x)->user->flags & FLAGS_SPOOFED)
 
 #define	SetDead(x)		((x)->flags |= FLAGS_DEADSOCK)
 #define	CBurst(x)		((x)->flags & FLAGS_CBURST)
@@ -238,6 +240,7 @@
 #define	SetEOB(x)		((x)->flags |= FLAGS_EOB)
 #define SetListenerInactive(x)	((x)->flags |= FLAGS_LISTENINACTIVE)
 #define SetKlineExempt(x)	((x)->user->flags |= FLAGS_EXEMPT)
+#define SetSpoofed(x)		((x)->user->flags |= FLAGS_SPOOFED)
 
 #define	DoingDNS(x)		((x)->flags & FLAGS_DOINGDNS)
 #define	DoingAuth(x)		((x)->flags & FLAGS_AUTH)
@@ -375,7 +378,7 @@
 #ifdef XLINE
 #define CFLAG_XEXEMPT		0x00080
 #endif
-
+#define CFLAG_SPOOFED		0x00200
 #define IsConfRestricted(x)	((x)->flags & CFLAG_RESTRICTED)
 #define IsConfRNoDNS(x)		((x)->flags & CFLAG_RNODNS)
 #define IsConfRNoIdent(x)	((x)->flags & CFLAG_RNOIDENT)
@@ -386,7 +389,7 @@
 #ifdef XLINE
 #define IsConfXlineExempt(x)	((x)->flags & CFLAG_XEXEMPT)
 #endif
-
+#define IsConfSpoofed(x)	((x)->flags & CFLAG_SPOOFED)
 #define PFLAG_DELAYED		0x00001
 #define PFLAG_SERVERONLY	0x00002
 
diff -ruN irc2.11.2p3.orig/common/support.c irc2.11.2p3/common/support.c
--- irc2.11.2p3.orig/common/support.c	2009-03-15 02:25:32.000000000 +0100
+++ irc2.11.2p3/common/support.c	2019-09-24 20:44:31.272196056 +0200
@@ -782,7 +782,7 @@
 char	*make_version(void)
 {
 	int ve, re, mi, dv, pl;
-	char ver[15];
+	char ver[32];
 
 	sscanf(PATCHLEVEL, "%2d%2d%2d%2d%2d", &ve, &re, &mi, &dv, &pl);
 	/* version & revision */
@@ -794,6 +794,9 @@
 		sprintf(ver + strlen(ver), "%c%d", DEVLEVEL, dv);
 	if (pl)	/* patchlevel */
 		sprintf(ver + strlen(ver), "p%d", pl);
+
+	strcat(ver, "+spoof");
+
 	return mystrdup(ver);
 }
 
diff -ruN irc2.11.2p3.orig/common/support.c.rej irc2.11.2p3/common/support.c.rej
--- irc2.11.2p3.orig/common/support.c.rej	1970-01-01 01:00:00.000000000 +0100
+++ irc2.11.2p3/common/support.c.rej	2019-09-23 23:08:12.728673430 +0200
@@ -0,0 +1,20 @@
+--- common/support.c   2008-07-15 17:24:29.000000000 +0200
++++ common/support.c        2009-12-09 19:17:03.590495354 +0100
+@@ -784,7 +784,7 @@
+ char   *make_version(void)
+ {
+        int ve, re, mi, dv, pl;
+-       char ver[50];
++       char ver[55];
+ 
+        sscanf(PATCHLEVEL, "%2d%2d%2d%2d%2d", &ve, &re, &mi, &dv, &pl);
+        /* version & revision */
+@@ -796,7 +796,7 @@
+                sprintf(ver + strlen(ver), "%c%d", DEVLEVEL, dv);
+        if (pl) /* patchlevel */
+                sprintf(ver + strlen(ver), "p%d", pl);
+-        strcat(ver,"+ssl+fic");
++        strcat(ver,"+ssl+fic+spoof");
+ 
+        return mystrdup(ver);
+ }
diff -ruN irc2.11.2p3.orig/ircd/s_conf.c irc2.11.2p3/ircd/s_conf.c
--- irc2.11.2p3.orig/ircd/s_conf.c	2010-08-11 19:16:51.000000000 +0200
+++ irc2.11.2p3/ircd/s_conf.c	2019-09-24 20:20:29.374024397 +0200
@@ -613,6 +613,12 @@
 		{
 			find_bounce(cptr, ConfClass(aconf), -1);
 		}
+		if(IsConfSpoofed(aconf))
+		{
+			strcpy(cptr->sockhost, aconf->name2);
+			strcpy(cptr->user->host, aconf->name2);
+			SetSpoofed(cptr);
+		}
 		break;
 	}
 	if (retval == -2)
@@ -1798,7 +1804,12 @@
 					iline_flags_parse(tmp3) :
 					pline_flags_parse(tmp3));
 			}
-
+			/* any spoofed host in this line? */
+			if(aconf->status & CONF_CLIENT && tmp4 && *tmp4)
+			{
+				DupString(aconf->name2, tmp4);
+				aconf->flags |= CFLAG_SPOOFED;
+			}	
 			/* trying to find exact conf line in already existing
 			 * conf, so we don't delete old one, just update it */
 			if (
diff -ruN irc2.11.2p3.orig/ircd/s_misc.c irc2.11.2p3/ircd/s_misc.c
--- irc2.11.2p3.orig/ircd/s_misc.c	2009-11-13 21:08:11.000000000 +0100
+++ irc2.11.2p3/ircd/s_misc.c	2019-09-24 23:00:34.767527724 +0200
@@ -196,7 +196,7 @@
 				(void)sprintf(nbuf, "%s[%.*s@%s]",
 					sptr->name, USERLEN,
 					(!(sptr->flags & FLAGS_GOTID)) ? "" :
-					sptr->auth, sptr->user ? sptr->user->sip :
+					sptr->auth, sptr->user ? get_client_ip(sptr) :
 #ifdef INET6 
 					      inetntop(AF_INET6,
 						       (char *)&sptr->ip,
@@ -247,6 +247,18 @@
 	return nbuf;
 }
 
+char	*get_client_ip(aClient *cptr)
+{
+	if(IsSpoofed(cptr))
+	{
+		return "255.255.255.255";
+	}
+	else
+	{
+		return cptr->user->sip;
+	}
+}
+
 /*
  * Form sockhost such that if the host is of form user@host, only the host
  * portion is copied.
diff -ruN irc2.11.2p3.orig/ircd/s_misc_ext.h irc2.11.2p3/ircd/s_misc_ext.h
--- irc2.11.2p3.orig/ircd/s_misc_ext.h	2004-03-14 18:46:00.000000000 +0100
+++ irc2.11.2p3/ircd/s_misc_ext.h	2019-09-24 22:06:51.119855303 +0200
@@ -42,6 +42,7 @@
 EXTERN int check_registered_service (aClient *sptr);
 EXTERN char *get_client_name (aClient *sptr, int showip);
 EXTERN char *get_client_host (aClient *cptr);
+EXTERN char *get_client_ip (aClient *cptr);
 EXTERN void get_sockhost (Reg aClient *cptr, Reg char *host);
 EXTERN char *my_name_for_link (char *name, Reg int count);
 EXTERN int mark_blind_servers (aClient *cptr, aClient *server);
diff -ruN irc2.11.2p3.orig/ircd/s_serv.c irc2.11.2p3/ircd/s_serv.c
--- irc2.11.2p3.orig/ircd/s_serv.c	2010-08-12 18:24:31.000000000 +0200
+++ irc2.11.2p3/ircd/s_serv.c	2019-09-24 22:26:14.957386472 +0200
@@ -1264,7 +1264,7 @@
 					   acptr->name, acptr->user->uid,
 					   acptr->user->username,
 					   acptr->user->host,
-					   acptr->user->sip,
+					   get_client_ip(acptr),
 					   (*buf) ? buf : "+", acptr->info);
 		    }
 		else if (IsService(acptr) &&
@@ -1783,6 +1783,9 @@
 #endif
 			else if ((tmp->status & CONF_CLIENT))
 			{
+				if(IsConfSpoofed(tmp))
+					continue;
+
 				sendto_one(sptr, replies[p[1]], ME, BadTo(to),
 					   c, host, (pass) ? "*" : null,
 					   name, port, get_conf_class(tmp),
@@ -2857,7 +2860,7 @@
 				IsAnOper(acptr) ? "Oper" : "User",
 				get_client_class(acptr),
 				acptr->name, acptr->user->username,
-				acptr->user->host, acptr->user->sip,
+				acptr->user->host, get_client_ip(acptr),
 #ifdef XLINE
 				acptr->user2, acptr->user3, 
 #else
@@ -2880,7 +2883,7 @@
 				IsAnOper(acptr) ? "Oper" : "User", 
 				get_client_class(acptr), 
 				acptr->name, acptr->user->username, 
-				acptr->user->host, acptr->user->sip,
+				acptr->user->host, get_client_ip(acptr),
 #ifdef XLINE
 				acptr->user2, acptr->user3, 
 #else
@@ -2916,7 +2919,7 @@
 			IsAnOper(acptr) ? "Oper" : "User", 
 			MyClient(acptr) ? get_client_class(acptr) : -1, 
 			acptr->name, acptr->user->username,
-			acptr->user->host, acptr->user->sip, 
+			acptr->user->host, get_client_ip(acptr), 
 #ifdef XLINE
 			MyClient(acptr) ? acptr->user2 : "-",
 			MyClient(acptr) ? acptr->user3 : "-",
diff -ruN irc2.11.2p3.orig/ircd/s_service.c irc2.11.2p3/ircd/s_service.c
--- irc2.11.2p3.orig/ircd/s_service.c	2010-08-12 03:08:02.000000000 +0200
+++ irc2.11.2p3/ircd/s_service.c	2019-09-24 22:27:26.652699997 +0200
@@ -180,7 +180,7 @@
 			sptr->user->uid,
 			(wants & SERVICE_WANT_USER) ? sptr->user->username : ".",
 			(wants & SERVICE_WANT_USER) ? sptr->user->host : ".",
-			(wants & SERVICE_WANT_USER) ? sptr->user->sip : ".",
+			(wants & SERVICE_WANT_USER) ? get_client_ip(sptr) : ".",
 			(wants & (SERVICE_WANT_UMODE|SERVICE_WANT_OPER)) ? umode : "+",
 			(wants & SERVICE_WANT_USER) ? sptr->info : "");
 	else
diff -ruN irc2.11.2p3.orig/ircd/s_user.c irc2.11.2p3/ircd/s_user.c
--- irc2.11.2p3.orig/ircd/s_user.c	2010-08-12 18:29:30.000000000 +0200
+++ irc2.11.2p3/ircd/s_user.c	2019-09-24 22:38:10.889590212 +0200
@@ -736,7 +736,7 @@
 			" :%s"
 # endif
 			, user->uid, nick, user->username,
-			user->host, user->sip
+			user->host, get_client_ip(sptr)
 # if (CLIENTS_CHANNEL_LEVEL & CCL_CONNINFO)
 #  ifdef XLINE
          , savedusername, sptr->user2, sptr->user3
@@ -806,7 +806,7 @@
 		sendto_one(acptr,
 				":%s UNICK %s %s %s %s %s %s :%s",
 				user->servp->sid, nick, user->uid,
-				user->username, user->host, user->sip,
+				user->username, user->host, get_client_ip(sptr),
 				(*buf) ? buf : "+", sptr->info);
 	}	/* for(my-leaf-servers) */
 #ifdef	USE_SERVICES
